src/context.ts

```diff
@@ -578,24 +566,53 @@ export class Context {
       disable_notification?: boolean
     }
   ) {
-    this.assert(this.message, 'forwardMessage')
+    const message = getMessageFromAnySource(this)
+    this.assert(message, 'forwardMessage')
     return this.telegram.forwardMessage(
       chatId,
-      this.message.chat.id,
-      this.message.message_id,
+      message.chat.id,
+      message.message_id,
       extra
     )
   }
 
   copyMessage(chatId: string | number, extra?: tt.ExtraCopyMessage) {
-    this.assert(this.message, 'copyMessage')
+    const message = getMessageFromAnySource(this)
+    this.assert(message, 'copyMessage')
     return this.telegram.copyMessage(
       chatId,
-      this.message.chat.id,
-      this.message.message_id,
+      message.chat.id,
+      message.message_id,
       extra
     )
   }
 }
 
 export default Context
+
+function getMessageFromAnySource(ctx: Context) {
+  return (
+    ctx.message ??
+    ctx.editedMessage ??
+    ctx.callbackQuery?.message ??
+    ctx.channelPost ??
+    ctx.editedChannelPost
+  )
+}
+
+function getQueryFromAnySource(ctx: Context) {
+  return (
+    ctx.callbackQuery ??
+    ctx.inlineQuery ??
+    ctx.shippingQuery ??
+    ctx.preCheckoutQuery
+  )
+}
```

Comment:
This function is only called by `getUserFromAnySource`, I'd rather inline this into `getUserFromAnySource` that, or into `ctx.from`.

Reply:
Preferably into `ctx.from`